Support tabs in @test definition

mbland opened the pull request:
Part of #20. The base of this is ported from sstephenson/bats#122 by @BenoitLefebvre to address sstephenson/bats#98. Added a couple of refinements on top of that original.

One of those refinements is to emit a warning if a test name is empty. If folks would rather, I can split that commit off into another PR.

sublimino commented on test/fixtures/bats/empty_name.bats:
```diff
@@ -0,0 +1,3 @@
+@test  { true; }
+@test '' { true; }
+@test '	 	 	 ' { true; }
```
Perhaps move "tabs and spaces" to individual tests and annotate/comment them for clarity?

jasonkarns replied:
If I can speak for @mbland here, I think this test is still just ensuring that a whitespace-only test name is still treated as "empty", so I think it still belongs in this set of fixtures.

sublimino commented on test/fixtures/bats/single_line.bats:
```diff
@@ -1,9 +1,9 @@
-@test "passing" { true; }
+@test "passing" {true;}
 
-@test "input redirection" { diff - <( echo hello ); } <<EOS
```
Any reason for removing spaces here and not in `empty_name.bats`? I prefer the readability of spaced, but it's another styleguide question to which only conformity is important!

jasonkarns replied:
Agree here. I'd rather the `tabs.bats` be changed to `whitespace.bats` and cover all the whitespace variants:
- no whitespace around single-line
- spaces around single-line
- tabs around single-line

- tabs before @test 
- tabs before test name
- tabs before test body
- spaces before @test 
- spaces before test name
- spaces before test body

mbland replied:
I've renamed the fixture to `whitespace.bats` and added a few extra cases, but didn't add extra cases for tabs vs. spaces, since the `[:space:]` character class matches both.

What _did_ happen is the new "extra whitespace around single-line test" case exposed a bug where the extra spaces after the closing quote were getting matched as part of the test case name. As a result, Bash would exit with a parse error when executing the preprocessed file, since the test name included the closing quote and its trailing leading to an unterminated quote parse error.

I've since updated `BATS_TEST_PATTERN` so that the name/description capture matches a non-`[:space:]` character last instead of first.

mbland commented:
FYI @sublimino points taken; will address later this evening.

jasonkarns commented:
@mbland I do think the warning might be better off in a separate PR. reason being... I'm beginning to see a theme around bats having its own "strict mode" so to speak. consider:

- warning about empty test names
- warning about duplicate test names
- bats itself being in or out-of UnofficialBashStrictMode
- bats "fixing" known bash gotchas (like `[[`)

Perhaps we should have a more comprehensive flag that enables bats "strict" mode (or maybe even better, defaults to strict mode and offer a --loose alternative). In this theoretical strict mode, empty and duplicate test names automatically fail; bats itself is placed in strict mode before the execution of every test (to ensure the sourced test doesn't disable it); and we automatically fix any bash gotchas that we know about?

mbland commented:
@jasonkarns and @sublimino I've rebased the branch, renamed `tabs.bats` to `whitespace.bats`, added new test cases, fixed a bug by updating `BATS_TEST_PATTERN` (see my reply to the review comment from earlier, or the commit message for a18c8f87e693ce079a9e40f49a22a15cbef52164), and peeled off the empty-name check for a future PR. PTAL.

Also, I like the idea of enabling a strict mode by default with a `--loose` or `BATS_DISABLE_STRICT_MODE` option for v1.0.0.

jasonkarns commented on libexec/bats:
```diff
@@ -77,7 +77,7 @@ abs_dirname '.' 'BATS_CWD'
 export BATS_PREFIX
 export BATS_CWD
-export BATS_TEST_PATTERN='^ *@test +(.+) +\{ *(.*)$'
+export BATS_TEST_PATTERN="^[[:space:]]*@test[[:space:]]+(.*[^[:space:]])[[:space:]]+\{(.*)\$"
```
Aren't the extra wrapping brackets around `[:space:]` unnecessary? (Except in the single negation instance?)

jasonkarns replied:
Apparently they _are_ required. TIL (I thought they operated just like the shortcut forms.)

mbland replied:
They're all required. Just tried locally to make sure, but the outside brackets just announce a set of characters to match, and `[:space:]` is special syntax announcing a character class within those outer brackets.

mbland commented:
@jasonkarns TIL about `[:blank:]`. How'd I miss that all these years? I've pushed a new commit with that change. PTAL?

Also, I don't think the Bash version would have an impact on the complexity of the regex at all. At least, off the top of my head, I can't think of why it would. Have you ideas?

jasonkarns commented:
@mbland 

> How'd I miss that all these years?

No judgement here. I just learned this morning when I was reading about the double brackets.



> complexity of the regex

I was mainly just thinking of character class shorthands, but I guess those aren't even in ERE...

mbland commented:
@jasonkarns Yeah, I think all the regex parsing is done by the standard C lib.

Just waiting for Travis to clear its macOS backlog before merging...
